### PR TITLE
Add `__array_ufunc__` and `__array__function__` to ArrayIndex

### LINKED ISSUE #95: ndarray(x) == NDIndex(x) and NDIndex(x) == ndarray(x) should return a consistent result
Currently, testing the equality of `NDIndex` vs `ndarray` objects changes depending on the order of the objects:

```python
>>> ndindex([0, 1]) == np.array([0, 1])
True
>>> np.array([0, 1]) == ndindex([0, 1])
array([False, False])
```

Ideally, the result should not change in this case, ie

```python
a = (ndindex([0, 1]) == np.array([0, 1]))
b = (np.array([0, 1]) == ndindex([0, 1]))
np.all(a == b) == True
```

### PR DESCRIPTION
**Pull request by telamonian**
Fixes #95

Adds appropriate definitions for the `__array_ufunc__` and `__array__function__` to the `ArrayIndex` class. This makes it so that if `a` is a `ndarray` and `b` is a `ArrayIndex`, both `a == b` and `b == a` get dispatched to `b.__eq__(a)`

### CODE REVIEW

**Comment by asmeurer** [comment]
You should update the `==` tests in test_ndindex.py.

What are the other consequences of this? I don't want array index types to be treatable as arrays (see https://quansight.github.io/ndindex/type-confusion.html#integerarray-and-booleanarray). Array operations should in general not work on them.

**Comment by asmeurer** [comment]
Oh I didn't see your comment on the issue. Given that, this does seem like the best option.

**Comment by asmeurer** [comment]
Also I just merged https://github.com/Quansight/ndindex/pull/93 which updates that test, so you'll need to update this on top of that.

File: ndindex/array.py [hunk 1]
```diff
@@ -14,6 +14,10 @@ class ArrayIndex(NDIndex):
     To subclass this, define the `dtype` attribute, as well as all the usual
     ndindex methods.
     """
+    __array_ufunc__ = None
```
Comment (asmeurer) [inline ndindex/array.py:17]: Add a comment here explaining why these are here.

**Comment by asmeurer** [comment]
This looks good. I would just like a comment explaining what this does. I guess it's more possible to override NumPy behavior than I thought. Now if only NumPy would allow `a[idx]` to be overridden so that we don't have to use `a[idx.raw]` for ndindex types.

**Comment by telamonian** [comment]
There's probably a NEP for that. Turns out there's a NEP for everything :)

**Comment by asmeurer** [comment]
I wonder if we should also at least define 

```py
def __array__(self):
    return self.array
```

Right now we have

```py
>>> a = np.array([True, False])
>>> idx = ndindex(a)
>>> np.array(idx)
array(BooleanArray([True, False]), dtype=object)
```

We could also make it give an error. 

```py
    def __array__(self):
        raise TypeError(f"Cannot convert {self.__class__.__name__} to an array. Use .array instead.")
```

I'm not sure which way is better.

**Comment by telamonian** [comment]
`ArrayIndex` instances are not arrays. So they shouldn't just transparently cast to arrays.

They're hardly rules to live by, but I've found the [zen of python](https://www.python.org/dev/peps/pep-0020/) to be good guidelines for building the simplest, most intuitable version of complex behaviors. Here I think the relevant parts are

"Explicit is better than implicit."

and

"There should be one-- and preferably only one --obvious way to do it."

So I think your second suggestion about raising a `TypeError` is good. But I'd go even further and say that the `.array` prop is superfluous and potentially confusing, and should be removed in favor of just `.raw`

**Comment by telamonian** [comment]
> This looks good. I would just like a comment explaining what this does. I guess it's more possible to override NumPy behavior than I thought. Now if only NumPy would allow `a[idx]` to be overridden so that we don't have to use `a[idx.raw]` for ndindex types.

Honestly this does seem like a good idea. Maybe we really should consider opening a NEP about ndarray.__getitem__ dispatch. I opened an issue for this at #97

**Comment by asmeurer** [comment]
Sorry for not looking into your PRs. The CI was broken on ndindex because of an upstream numpy issue, but that's been resolved now that the 1.20 prerelease is out. I've pushed a commit here addressing my comment, and I will merge once the tests pass.

**Pull request merged**